### PR TITLE
add command to sync plan data from stripe, doing testing

### PR DESCRIPTION
**Pull request by anamarn**
**Solves**
[ https://github.com/twentyhq/private-issues/issues/211 ]

**TLDR:**

Add a command that fetches the plans product, meters and price in stripe and whrites it to the DataBase. For now it fetches only active products. 

**In order to test**

- Set IS_BILLING_ENABLED=true
- Run `npx nx database:reset twenty-server` if you don't have the billing tables in your data base schema 
-  run `npx nx run twenty-server:command billing:sync-plans-data -v`

### CODE REVIEW

File: packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    passedParams: string[],
+    options: BaseCommandOptions,
+  ): Promise<void> {
+    const billingMeters = await this.stripeService.getAllMeters();
+
+    await this.upsertMetersRepositoryData(billingMeters, options);
+
+    const billingProducts = await this.stripeService.getAllProducts();
+
+    const billingProductsIds = await Promise.all(
```
Comment (Weiko) [inline packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts:87]: Do we expect a lot of records in that one? If so, Promise.all could benefit from using batching here

File: packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+          this.logger.log(
+            `Product: ${product.id} purposefully not inserted, invalid metadata format: ${JSON.stringify(
+              product.metadata,
+            )}`,
+          );
+        }
+      }),
+    );
+
+    const billingPrices = await Promise.all(
```
Comment (Weiko) [inline packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts:103]: same here, could make a util

Comment (Weiko) [reply packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts:103]: Actually, can't we do all the operations in the same map()? Looks like you don't need to split those

Comment (Weiko) [reply packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts:103]: Haven't tested this but that's the idea
```typescript

    const billingPrices = await Promise.all(
      billingProducts.map(async (product) => {
        if (!isStripeValidProductMetadata(product.metadata)) {
          this.logger.log(
            `Product: ${product.id} purposefully not inserted, invalid metadata format: ${JSON.stringify(
              product.metadata,
            )}`,
          );

          return [];
        }

        await this.upsertProductRepositoryData(product, options);

        const prices = await this.stripeService.getPricesByProductId(
          product.id,
        );

        this.logger.log(
          `${prices.length} prices found for product: ${product.id}`,
        );

        return prices;
      }),
    );
   ```

File: packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+          );
+
+          return prices;
+        }
+
+        return [];
+      }),
+    );
+
+    const transformedPrices = billingPrices
```
Comment (Weiko) [inline packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts:120]: FYI you could use flatMap here
```typescript
billingPrices.flatMap(transformStripePriceDataToPriceRepositoryData);
```

File: packages/twenty-server/src/engine/core-modules/billing/utils/transform-stripe-price-data-to-price-repository-data.util.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      data.currency_options === null ? undefined : data.currency_options,
+    tiers: data.tiers === null ? undefined : data.tiers,
+    tiersMode: data.tiers_mode
+      ? getBillingPriceTiersMode(data.tiers_mode)
+      : undefined,
+    recurring: data.recurring === null ? undefined : data.recurring,
+  };
+};
+
+const getTaxBehavior = (data: Stripe.Price.TaxBehavior) => {
```
Comment (Weiko) [inline packages/twenty-server/src/engine/core-modules/billing/utils/transform-stripe-price-data-to-price-repository-data.util.ts:46]: Not a strong opinion but other way to do this is to create a map
```typescript
const TAX_BEHAVIOR_MAP: Record<Stripe.Price.TaxBehavior, BillingPriceTaxBehavior> = {
  exclusive: BillingPriceTaxBehavior.EXCLUSIVE,
  inclusive: BillingPriceTaxBehavior.INCLUSIVE,
  unspecified: BillingPriceTaxBehavior.UNSPECIFIED,
};

const getTaxBehavior = (data: Stripe.Price.TaxBehavior) => TAX_BEHAVIOR_MAP[data];
```
up to you

**Review by Weiko — COMMENTED** [review]
Left a few comments, code LGTM, good job!
Haven't tested the command yet

File: packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts (hunk 1 above)
Comment (anamarn) [reply packages/twenty-server/src/engine/core-modules/billing/commands/billing-sync-plans-data.command.ts:87]: I estimate that we will have between 10 and 100  in the course of Twenty's life (realistically we will have approx 15). Is that considered enough records to consider the use of a batch?

**Review by FelixMalfait — APPROVED** [review]

**Pull request merged**